File: test/core/test_is_dir.f90
program test_is_dir

use, intrinsic :: iso_fortran_env, only: stderr=>error_unit
use filesystem

implicit none

valgrind : block

character(:), allocatable :: s1, s2
integer :: i


if(is_dir("")) error stop "is_dir empty should be false"

s1 = get_cwd()
if(.not. is_dir(s1)) then
  write(stderr, '(a)') "is_dir(get_cwd()) failed: " // s1
  error stop
endif

if(is_windows()) then
  s1 = root(s1)
  print '(3A,i0)', "root(get_cwd()) = ", s1, " length = ", len_trim(s1)
  if(.not. is_dir(s1)) error stop "is_dir('" // s1 // "') failed"
else
  if(.not. is_dir("/")) error stop "is_dir('/') failed"
endif

if(.not. is_dir(".")) error stop "did not detect '.' as directory"
if(is_file(".")) error stop "detected '.' as file"
call assert_is_dir(".")

if(.not. is_readable(".")) error stop "%is_readable failed on '.'"

allocate(character(max_path()) :: s2)
call get_command_argument(0, s2, status=i)
if(i /= 0) error stop "get_command_argument(0) failed"

if (is_dir(s2)) error stop "detected file as directory"

if(is_dir("not-exist-dir")) error stop "not-exist-dir should not exist"

end block valgrind

end program

File: test/core/test_is_dir.cpp
#include <iostream>
#include <cstdlib>

#include "ffilesystem.h"
#include "ffilesystem_test.h"


int main(int argc, char* argv[])
{

  std::string file = (argc > 1) ? argv[1] : argv[0];

  if (fs_is_dir(""))
    err("is_dir empty should be false");

  std::string s1 = fs_get_cwd();
  if (!fs_is_dir(s1))
    err("is_dir(get_cwd()) failed: " + s1);

  if (!fs_is_dir("."))
    err("did not detect '.' as directory");

  if (fs_is_file("."))
    err("detected '.' as file");

  if(!fs_is_dir("."))
    err("is_dir failed on '.'");

  if (!fs_exists(".") || !fs_is_dir("."))
    err("is_readable failed on '.'");

  std::string_view s2(file);
  if (fs_is_dir(s2))
    err("detected file as directory");

  if (fs_is_dir("not-exist-dir"))
    err("not-exist-dir should not exist");

  std::string fn = "日本語_is_dir";

  if (!fs_mkdir(fn))
    err("fs_mkdir " + fn);

  if (!fs_exists(fn))
    err("fs_exists " + fn);

  if (!fs_is_dir(fn))
    err("fs_is_dir " + fn);

  ok_msg("is_dir C++");

  return EXIT_SUCCESS;
}
